src/polkit/polkitunixprocess.c

```diff
@@ -1131,6 +1362,445 @@ get_start_time_for_pid (pid_t    pid,
+    }
+
+  tokens = g_strsplit (p, " ", 0);
+
+  num_tokens = g_strv_length (tokens);
+
+  if (num_tokens < 5)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+
+  /*
+   * The string can have negative representation, but it's actually an unsigned int,
+   * cast it to convert it
+   */
+  tty_nr = (guint) strtoll (tokens[4], &endp, 10);
+  if (endp == tokens[4])
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+
+  if (tty_nr == 0)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "No TTY associated with process %d",
+                   pid);
+      goto out;
+    }
+
+  /* Now do some sanity checks to ensure we are really looking at a TTY. */
+
+  /*
+   * As per the manpage, the minor device number is contained in the
+   * combination of bits 31 to 20 and 7 to 0, the major device number is in
+   * bits 15 to 8.
+   */
+  major = (tty_nr & 0xff00) >> 8;
+  minor = ((tty_nr & 0xfff00000) >> 12) | (tty_nr & 0xff);
+
+  tty_name = g_strdup_printf ("/dev/pts/%lu", minor);
+
+  tty_fd = open (tty_name, O_RDONLY|O_CLOEXEC);
+  if (tty_fd < 0)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error opening TTY %s for process %d: %s",
+                   tty_name,
+                   pid,
+                   g_strerror (errno));
+      goto out;
+    }
+
+  if (fstat (tty_fd, &st) < 0)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error getting TTY name for process %d: %s",
+                   pid,
+                   g_strerror (errno));
+      goto out;
+    }
+
+  if (!S_ISCHR (st.st_mode))
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "TTY name for process %d is not a character device",
+                   pid);
+      goto out;
+    }
+
+  if (major != major (st.st_rdev) || minor != minor (st.st_rdev))
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "TTY device number mismatch for process %d",
+                   pid);
+      goto out;
+    }
+
+  if (fstatfs (tty_fd, &stfs) < 0)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error getting TTY filesystem information for process %d: %s",
+                   pid,
+                   g_strerror (errno));
+      goto out;
+    }
+
+  if (stfs.f_type != 0x1cd1) /* 0x1cd1 is the magic number for devpts */
```

Comment:
you opened the device via /dev/pts/, hence this check is unnecessary, you know it's below /dev/pts...

Reply:
You told me to do this:

> if you want to properly check if something is a pty on linux, better look if **backing fs is devpts**, and if so check how it reacts to TIOCGPKT

https://github.com/polkit-org/polkit/pull/533#discussion_r1890338449

Reply:
if you have an open fd, sure. but if you open the fd anew, and you already went through the path to get it then why check this again?